File: Data Structures & Algorithms/word-break/submission-6.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = {len(s):True}
        def dfs(i: int) -> bool:
            if i in dp:
                return dp[i]
            if i == len(s):
                return True
            for w in wordDict:
                l = len(w)
                if i + l <= len(s) and s[i:i + l] == w:
                    if dfs(i + l):
                        dp[i] = True
                        return True
            dp[i] = False
            return False
        return dfs(0)
            

File: Data Structures & Algorithms/combination-target-sum/submission-6.cpp
class Solution {
public:
    vector<vector<int>> result;
    int max_len;
    int t;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        max_len = (int)nums.size();
        t = target;
        vector<int> curr = {};
        dfs(0, 0, curr, nums);
        return result;
    }
    void dfs(int i, int curr_sum, vector<int>& curr, vector<int>& nums) {
        if (curr_sum == t) {
            result.push_back(curr);
            return;
        }
        if (i >= max_len or curr_sum > t) {
            return;
        }
        curr.push_back(nums[i]);
        dfs(i, curr_sum + nums[i], curr, nums);
        curr.pop_back();
        dfs(i + 1, curr_sum, curr, nums);
    }
};
